### ISSUE TITLE
SuperWebSocket - Host in parallel via WinHTTP/HttpListener

### ISSUE THREAD

**Issue by Danielku15** [kerryjiang/SuperSocket#148]
We have made very good experience with SuperWebSocket (Performance and Architecture) and would like to continue using it. One of the problems we face is that SuperWebSocket needs to be operated on a dedicated port beside 80/443 if you also run a IIS application on your server. 

Microsofts answer to this problem is WinHTTP which allows applications to accept web based connections on same ports with different bindings. Like: 

- http://myserver.com/myapp -> Goes to an IIS web application
- http://myserver.com/myapi -> Goes to a self hosted ASP.net Web API

Also HTTPs is configured via WinHTTP (netsh http) and managed fully by the operating system. 

It would be nice if SuperWebSocket supports additionally to the raw socket server also to accept WebSocket connections via HttpListener class. This way we could keep using the SuperSocket architecture but in parallel utilize the shared ports. 

Especially in environments with load balancers and firewalls which are common in Cloud hosting environments you would like to stick to 80/443 for your web services. 

A small drawback of WinHTTP is that it only supports WebSockets on Windows 8 (or Server 2012).

**Comment by kerryjiang** [comment]
HttpListener would be completely different design.

Yes, could you just use something like proxy or load balancer for now?

**Comment by Danielku15** [comment]
Yes, the HttpListener requires quite a bit of redesign in the WebSocketServer as the underlying context already manages the handshake and data encoding. 

But I think it could be quite well be integrated into the current architecture:

* A new ISocketServer implementation (e.g. inheriting from the currently used AsyncSocketServer) could additionally fire up a HttpListener. This SocketServer is then actually a "dual" socket listener. If a client connects via HttpListener it sets up a session just like on the raw TCP socket. 
* `class WebSocketServer : WebSocketServer<WebSocketSession>` should become `class WebSocketServer : WebSocketServer<IWebSocketSession>`. This likely requires some rework as the session type need to be currently instanciatable by new(). I would rather make the CreateAppSession fully abstract and require the subclasses to instanciate them as needed. 
* In the CreateAppSession() of the WebSocketServer it is then detected whether a classical raw TCP socket which requires a handshake or a "HttpListenerSocketSession" is passed. Based on this specific implementations for the IWebSocketSession are provided. 
* The ProcotolProcessor logic (handshake, package encoding etc.) which currently lives in the WebSocketServer, could move down to the WebSocketSession as those are anyhow session specific things.  

A Proxy/LoadBalancer solution sounds hacky to me. It would be a in-process TCP tunnel. The other solution would rather be to build another abstraction layer between SuperWebSocket and our actual application logic. Then fire up SuperWebSocket as "legacy" server and additionally a HttpListener (maybe using SuperSocket) as new server. 

But as SuperSocket already has this nice abstraction of sessions it would be nice to have this abstraction 1 level deeper in the WebSocketServer implementation.

**Comment by Danielku15** [comment]
I am currently trying to implement a parallel HttpListener that utilizes the SuperSocket session infrastructure as good as possible. 

The biggest challenge currently seems that the SocketEngine implementations are internal and I need to wrap most things. Is there any special reason for that?

It would be nice to simply provide a `ISocketListener` implementation which wraps the `HttpListener` class. The 2 obstacles I have are that 
* everything in this area is internal and you cannot build your own logic on top. -> Is it feasable that you make the SocketEngine with it's interfaces and implementations public?
* the ListenerInfo does not have a custom metadata dictionary which would allow me to decide what listener to create in the `CreateListener` method. (`TcpAsyncSocketListener` or `HttpListenerSocketListener`) -> It would be nice if `IListenerConfig` interface would have a Dictionary<string,object> where you can add custom details which are then also copied to the ListenerInfo.
